File: android/src/main/java/com/oktareactnative/HttpClientImpl.java
/*
 * Copyright (c) 2019, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

package com.oktareactnative;

import static com.okta.oidc.net.ConnectionParameters.USER_AGENT;

import android.net.Uri;

import androidx.annotation.NonNull;

import com.okta.oidc.BuildConfig;
import com.okta.oidc.net.ConnectionParameters;
import com.okta.oidc.net.OktaHttpClient;

import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class HttpClientImpl implements OktaHttpClient {
    private final String userAgentTemplate;
    private final int connectTimeoutMs;
    private final int readTimeoutMs;
    protected static OkHttpClient sOkHttpClient;
    protected volatile Call mCall;
    protected Response mResponse;
    protected Exception mException;

    HttpClientImpl(String userAgentTemplate, int connectTimeoutMs, int readTimeoutMs) {
        this.userAgentTemplate = userAgentTemplate;
        this.connectTimeoutMs = connectTimeoutMs;
        this.readTimeoutMs = readTimeoutMs;
    }

    private String getUserAgent() {
        String sdkVersion = "okta-oidc-android/" + BuildConfig.VERSION_NAME;
        return userAgentTemplate.replace("$UPSTREAM_SDK", sdkVersion);
    }

    protected Request buildRequest(Uri uri, ConnectionParameters param) {
        if (sOkHttpClient == null) {
            sOkHttpClient = new OkHttpClient.Builder()
                    .connectTimeout(connectTimeoutMs, TimeUnit.MILLISECONDS)
                    .readTimeout(readTimeoutMs, TimeUnit.MILLISECONDS)
                    .build();
        }
        Request.Builder requestBuilder = new Request.Builder().url(uri.toString());
        requestBuilder.addHeader(USER_AGENT, getUserAgent());
        for (Map.Entry<String, String> headerEntry : param.requestProperties().entrySet()) {
            String key = headerEntry.getKey();
            requestBuilder.addHeader(key, headerEntry.getValue());
        }
        if (param.requestMethod() == ConnectionParameters.RequestMethod.GET) {
            requestBuilder = requestBuilder.get();
        } else {
            Map<String, String> postParameters = param.postParameters();
            if (postParameters != null) {
                FormBody.Builder formBuilder = new FormBody.Builder();
                for (Map.Entry<String, String> postEntry : postParameters.entrySet()) {
                    String key = postEntry.getKey();
                    formBuilder.add(key, postEntry.getValue());
                }
                RequestBody formBody = formBuilder.build();
                requestBuilder.post(formBody);
            } else {
                requestBuilder.post(RequestBody.create(null, ""));
            }
        }
        return requestBuilder.build();
    }

    @Override
    public InputStream connect(@NonNull Uri uri, @NonNull ConnectionParameters params)
            throws Exception {
        Request request = buildRequest(uri, params);
        mCall = sOkHttpClient.newCall(request);
        final CountDownLatch latch = new CountDownLatch(1);
        mCall.enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                mException = e;
                latch.countDown();
            }

            @Override
            public void onResponse(Call call, Response response) {
                mResponse = response;
                latch.countDown();
            }
        });
        latch.await();
        if (mException != null) {
            throw mException;
        }
        if (mResponse != null && mResponse.body() != null) {
            return mResponse.body().byteStream();
        }
        return null;
    }

    @Override
    public void cleanUp() {
        //NO-OP
    }

    @Override
    public void cancel() {
        if (mCall != null) {
            mCall.cancel();
        }
    }

    @Override
    public Map<String, List<String>> getHeaderFields() {
        if (mResponse != null) {
            return mResponse.headers().toMultimap();
        }
        return null;
    }

    @Override
    public String getHeader(String header) {
        if (mResponse != null) {
            return mResponse.header(header);
        }
        return null;
    }

    @Override
    public int getResponseCode() throws IOException {
        if (mResponse != null) {
            return mResponse.code();
        }
        return -1;
    }

    @Override
    public int getContentLength() {
        if (mResponse != null && mResponse.body() != null) {
            return (int) mResponse.body().contentLength();
        }
        return -1;
    }

    @Override
    public String getResponseMessage() throws IOException {
        if (mResponse != null) {
            return mResponse.message();
        }
        return null;
    }
}

File: android/src/main/java/com/oktareactnative/HttpClientImpl.kt
package com.oktareactnative

import android.annotation.SuppressLint
import android.net.Uri
import android.os.Build
import com.okta.oidc.BuildConfig
import com.okta.oidc.net.ConnectionParameters
import com.okta.oidc.net.ConnectionParameters.USER_AGENT
import com.okta.oidc.net.OktaHttpClient
import com.okta.oidc.net.request.TLSSocketFactory
import java.io.DataOutputStream
import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import java.security.KeyManagementException
import java.security.NoSuchAlgorithmException
import javax.net.ssl.HttpsURLConnection


class HttpClientImpl internal constructor(
  private val userAgentTemplate: String,
  private val connectTimeoutMs: Int,
  private val readTimeoutMs: Int
) :
  OktaHttpClient {
  var urlConnection: HttpURLConnection? = null
    private set

  /*
     * TLS v1.1, v1.2 in Android supports starting from API 16.
     * But it enabled by default starting from API 20.
     * This method enable these TLS versions on API < 20.
     * */
  @SuppressLint("RestrictedApi")
  private fun enableTlsV1_2(urlConnection: HttpURLConnection) {
    try {
      (urlConnection as HttpsURLConnection)
        .setSSLSocketFactory(TLSSocketFactory())
    } catch (e: NoSuchAlgorithmException) {
      throw RuntimeException("Cannot create SSLContext.", e)
    } catch (e: KeyManagementException) {
      throw RuntimeException("Cannot create SSLContext.", e)
    }
  }

  private val userAgent: String
    private get() {
      val sdkVersion = "okta-oidc-android/" + BuildConfig.VERSION_NAME
      return userAgentTemplate.replace("\$UPSTREAM_SDK", sdkVersion)
    }

  @Throws(IOException::class)
  protected fun openConnection(url: URL, params: ConnectionParameters): HttpURLConnection {
    val conn = url.openConnection() as HttpURLConnection
    if (urlConnection is HttpsURLConnection &&
      Build.VERSION.SDK_INT <= Build.VERSION_CODES.LOLLIPOP
    ) {
      enableTlsV1_2(urlConnection as HttpsURLConnection)
    }
    conn.setConnectTimeout(connectTimeoutMs)
    conn.setReadTimeout(readTimeoutMs)
    conn.instanceFollowRedirects = false
    val requestProperties = params.requestProperties()
    val userAgent = userAgent
    requestProperties[USER_AGENT] = userAgent
    for (property in requestProperties.keys) {
      conn.setRequestProperty(property, requestProperties[property])
    }
    val requestMethod = params.requestMethod()
    val postParameters = params.postParameters()
    conn.setRequestMethod(requestMethod.name)
    if (requestMethod == ConnectionParameters.RequestMethod.GET) {
      conn.setDoInput(true)
    } else if (requestMethod == ConnectionParameters.RequestMethod.POST) {
      conn.setDoOutput(true)
      if (postParameters != null && !postParameters.isEmpty()) {
        val out = DataOutputStream(conn.outputStream)
        out.write(params.encodedPostParameters)
        out.close()
      }
    }
    return conn
  }

  @Throws(Exception::class)
  override fun connect(uri: Uri, params: ConnectionParameters): InputStream? {
    urlConnection = openConnection(URL(uri.toString()), params)
    urlConnection!!.connect()
    return try {
      urlConnection!!.inputStream
    } catch (e: IOException) {
      urlConnection!!.errorStream
    }
  }

  override fun cleanUp() {
    urlConnection = null
  }

  override fun cancel() {
    if (urlConnection != null) {
      urlConnection!!.disconnect()
    }
  }

  override fun getHeaderFields(): Map<String, List<String>>? {
    return if (urlConnection != null) {
      urlConnection!!.headerFields
    } else null
  }

  override fun getHeader(header: String): String? {
    return if (urlConnection != null) {
      urlConnection!!.getHeaderField(header)
    } else null
  }

  @Throws(IOException::class)
  override fun getResponseCode(): Int {
    return if (urlConnection != null) {
      urlConnection!!.getResponseCode()
    } else -1
  }

  override fun getContentLength(): Int {
    return if (urlConnection != null) {
      urlConnection!!.getContentLength()
    } else -1
  }

  @Throws(IOException::class)
  override fun getResponseMessage(): String? {
    return if (urlConnection != null) {
      urlConnection!!.getResponseMessage()
    } else null
  }
}

